### DISCUSSION TITLE
Add `app.storage.tab` or similar which allows storing data per browser tab

### DISCUSSION THREAD

**Discussion by rodja** [zauberzeug/nicegui#1308] [Ideas / Feature Requests]
In https://github.com/zauberzeug/nicegui/pull/967#issuecomment-1661176431, @amks1 asked for another dict in `app.storage` which can hold data for an individual browser tab in contrast to the `app.storage.browser` which is a persistence available for a user session (across multiple tabs).

**Comment by rodja** [comment]
I'm not sure if this is really needed and how it should behave. There are two options:

1. If we provide a storage dict for each individual page it is no real "storage" (eg. persistent) because the next page would have its own storage. Users can do this already by creating a dict inside a page-builder.
2. Somehow storing by "tab". Adding data in one page and then moving to another inside the same tab would make the data accessible by the later pages. How would we implement this? And in what kind of scenarios would that be beneficial?

**Comment by amks1** [comment]
> 2. Somehow storing by "tab". Adding data in one page and then moving to another inside the same tab would make the data accessible by the later pages. How would we implement this? And in what kind of scenarios would that be beneficial?

Yes this is what I had started the [initial discussion](https://github.com/zauberzeug/nicegui/discussions/373) for 🙂. Each tab has to use its own websocket connection with a unique SSID (I'm no expert - but this is how it works on my webapp), so keeping a dictionary for each SSID should be doable. 

The use case is that a single user can open multiple instances of the same app without conflicts. At the server side we can spread out our app across multiple functions and use this dict anywhere without having to pass a bunch of variables around.

**Comment by rodja** [comment]
Good thinking, @amks1. We should do some experiments of how stable the socket id (SSID) is. For example I'm not sure if a new socket id is created if we reconnect due to short network loss.

I needed some time but now think that there are actual use cases for this. For example a questionnaire with forms spreading over multiple pages. It would be neat to have a convenient place to put the data, so that two tabs do not interfere. The main downside is, that it is no real storage (eg. persistent). The socket ids are randomly generated on connect.

**Comment by rodja** [comment]
There was a very interesting and fruitful discussion over on [Discord](https://discord.com/channels/1089836369431498784/1138784973625757706): Handling login credentials with `app.storage.tab` would be more secure and might justify re-login for every tab or server reboot depending on the use case. For example logging into a bank account or accessing a password manager.

**Comment by rodja** [comment]
In https://github.com/zauberzeug/nicegui/discussions/2378 @sklibanov312 found a possible approach using the Session Storage of the browser.

**Reply by sklibanov312** [reply]
Let's talk about the API we'd want to expose to the user. 

Question 1:
You can see in #2378 that I propose retrieving (and setting) several keys in one round-trip to the client. That's not quite the same as exposing a dictionary-like interface like we do elsewhere in storage.py. What's more important for nicegui? Speed or internal API consistency? I think speed wins this one, although maybe that can be optionally wrapped by a dict-like accessor. Any advice here?

My first preference would be e.g. 

```py
await app.storage.tab.setMany({'key1': 'value1', 'key2': 'value2'})

result = await app.storage.tab.getMany(['key1', 'key2'])
# result is a dict here. If key2 is absent, for example, it's omitted from result
```
For that to work we'd need to be sure that the Client is available in app.storage... Is that always the case? We'd need to add an optional timeout argument as well. Do we accept that awaiting a client at the start of setMany and getMany is reasonable?

Question 2:
Also it looks like for cookie-based storage, SessionMiddleware chooses to encrypt the cookie. Do we need to encrypt or sign session-local data that we store in the client? My application does not have this requirement, since all I plan to store is the user's own recent inputs. Is that good enough? Or do we need encryption of per-tab storage?

Question 3:
Do we require data binding support? If we do, can we assume that the nicegui server is the only thing changing the client's sessionStorage? Or do we need to listen for sessionStorage events (https://developer.mozilla.org/en-US/docs/Web/API/Window/storage_event) in the .js? Is that something that can come in later, or do we need that in this first PR also?

**Comment by rodja** [comment]
Thanks for taking up the ball, @sklibanov312. I agree, performance should be considered. But maybe we can have both: efficient save/storage and the same API as the other storages. We already have the concept of a accumulation/merging messages before sending them in `outbox.py`.

**Comment by rodja** [comment]
I've played around with different concepts but have not come up with a nice solution. Here some pieces I've thought about:

- using the same (or very similar) API as the other storages is preferable (due to consistency and the ability to use bindings)
- syncing the whole (flat) key list on every dict change is not very efficient but could be worth given the requirement above
- but how would we initialize the `ObservableDict` representing the tab-based storage in Python? The data can only become available after the client connection has been established (eg. when js can be executed).

The API suggestion from @sklibanov312 (`await app.storage.tab.getMany(['key1', 'key2'])`) would not have such a problem because of the awaited access. But this approach can not be used with bindings...

**Reply by Alyxion** [reply]
Thanks for the link @rodja.

I think its crucial to point out that involving the browser and trying to transfer data between client and browser would result in a far worse behavior than it is currently implemented in Streamlit.

There are several uses cases we have internally - and I am pretty sure other devs / Data Scientists have them too - where we are storing "living objects" thus arbitrary Python objects, e.g. Machine Learning models generated on a **per user basis** (in Streamlit effectively a "per browser tab") basis. Its impossible to transfer them to the browser and/or receive it as they can be up to multiple **hundred megabytes per user** and of course we would not want to transfer our know how/the model weights anyway.

Also its important to note that like @amks1 pointed out there are several use cases where its even mandatory that the data is discarded (e.g. user credentials) when a browser tab is closed, thus its even intended behavior that the data shall **not** be persisted.

Other than that I've started implementing a "per-session storage" which already works pretty well in conjunction with the single_page example:
https://github.com/Alyxion/nicegui/blob/feature/client_data/nicegui/storage.py

It does not work though yet the way it is intended to do when switching between "real" pages, see new login example:
https://github.com/Alyxion/nicegui/blob/feature/client_data/examples/session_storage/main.py

Making good progress though and optimistic it works at the end of the week for the example above.

One hurdle I will likely though **not** be able to solve so easily is the "reconnection" scenario where a connection was interrupted. This will require a smarter mechanism to retain the data for a given amount of seconds (until the reconnection definitely timed out) and reassigning it to the same session when its re-established.

**Comment by rodja** [comment]
@Alyxion I look forward to see your solution.

**Reply by Alyxion** [reply]
Thanks. Still got a working version running yesterday evening and created a pull request draft:

https://github.com/zauberzeug/nicegui/pull/2811

**Comment by rodja** [comment]
Thanks to @Alyxion we found a nice API for `app.storage.tab`. See https://github.com/zauberzeug/nicegui/pull/2837.

**Reply by falkoschindler** [reply]
This PR is part of the 1.4.21 release.
